Part III: Soccer Data

Introductory - Intermediate level SQL

Please complete this exercise using sqlite3 and Jupyter notebook.

Download the SQLite database and load in your notebook using the sqlite3 library.


In [2]:
import sqlite3

In [27]:
conn = sqlite3.connect('database-2.sqlite')

In [28]:
c = conn.cursor()

In [29]:
c.execute("SELECT name FROM sqlite_master WHERE type='table';")
print(c.fetchall())

[('sqlite_sequence',), ('Player_Attributes',), ('Player',), ('Match',), ('League',), ('Country',), ('Team',), ('Team_Attributes',)]


In [36]:
c.execute('SELECT * FROM sqlite_sequence LIMIT 5')
c.fetchall()

[('Team', 103916),
 ('Country', 51958),
 ('League', 51958),
 ('Match', 51958),
 ('Player', 11075)]

In [37]:
c.description

(('name', None, None, None, None, None, None),
 ('seq', None, None, None, None, None, None))

In [38]:
c.execute('SELECT * FROM Player_Attributes LIMIT 5')
c.fetchall()

[(1,
  218353,
  505942,
  '2016-02-18 00:00:00',
  67,
  71,
  'right',
  'medium',
  'medium',
  49,
  44,
  71,
  61,
  44,
  51,
  45,
  39,
  64,
  49,
  60,
  64,
  59,
  47,
  65,
  55,
  58,
  54,
  76,
  35,
  71,
  70,
  45,
  54,
  48,
  65,
  69,
  69,
  6,
  11,
  10,
  8,
  8),
 (2,
  218353,
  505942,
  '2015-11-19 00:00:00',
  67,
  71,
  'right',
  'medium',
  'medium',
  49,
  44,
  71,
  61,
  44,
  51,
  45,
  39,
  64,
  49,
  60,
  64,
  59,
  47,
  65,
  55,
  58,
  54,
  76,
  35,
  71,
  70,
  45,
  54,
  48,
  65,
  69,
  69,
  6,
  11,
  10,
  8,
  8),
 (3,
  218353,
  505942,
  '2015-09-21 00:00:00',
  62,
  66,
  'right',
  'medium',
  'medium',
  49,
  44,
  71,
  61,
  44,
  51,
  45,
  39,
  64,
  49,
  60,
  64,
  59,
  47,
  65,
  55,
  58,
  54,
  76,
  35,
  63,
  41,
  45,
  54,
  48,
  65,
  66,
  69,
  6,
  11,
  10,
  8,
  8),
 (4,
  218353,
  505942,
  '2015-03-20 00:00:00',
  61,
  65,
  'right',
  'medium',
  'medium',
  48,
  43,
  70,
  60,


In [39]:
c.description

(('id', None, None, None, None, None, None),
 ('player_fifa_api_id', None, None, None, None, None, None),
 ('player_api_id', None, None, None, None, None, None),
 ('date', None, None, None, None, None, None),
 ('overall_rating', None, None, None, None, None, None),
 ('potential', None, None, None, None, None, None),
 ('preferred_foot', None, None, None, None, None, None),
 ('attacking_work_rate', None, None, None, None, None, None),
 ('defensive_work_rate', None, None, None, None, None, None),
 ('crossing', None, None, None, None, None, None),
 ('finishing', None, None, None, None, None, None),
 ('heading_accuracy', None, None, None, None, None, None),
 ('short_passing', None, None, None, None, None, None),
 ('volleys', None, None, None, None, None, None),
 ('dribbling', None, None, None, None, None, None),
 ('curve', None, None, None, None, None, None),
 ('free_kick_accuracy', None, None, None, None, None, None),
 ('long_passing', None, None, None, None, None, None),
 ('ball_control',

In [40]:
c.execute('SELECT * FROM Player LIMIT 5')
c.fetchall()

[(1, 505942, 'Aaron Appindangoye', 218353, '1992-02-29 00:00:00', 182.88, 187),
 (2, 155782, 'Aaron Cresswell', 189615, '1989-12-15 00:00:00', 170.18, 146),
 (3, 162549, 'Aaron Doran', 186170, '1991-05-13 00:00:00', 170.18, 163),
 (4, 30572, 'Aaron Galindo', 140161, '1982-05-08 00:00:00', 182.88, 198),
 (5, 23780, 'Aaron Hughes', 17725, '1979-11-08 00:00:00', 182.88, 154)]

In [41]:
c.description

(('id', None, None, None, None, None, None),
 ('player_api_id', None, None, None, None, None, None),
 ('player_name', None, None, None, None, None, None),
 ('player_fifa_api_id', None, None, None, None, None, None),
 ('birthday', None, None, None, None, None, None),
 ('height', None, None, None, None, None, None),
 ('weight', None, None, None, None, None, None))

In [42]:
c.execute('SELECT * FROM Match LIMIT 5')
c.fetchall()

[(1,
  1,
  1,
  '2008/2009',
  1,
  '2008-08-17 00:00:00',
  492473,
  9987,
  9993,
  1,
  1,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  1.73,
  3.4,
  5,
  1.75,
  3.35,
  4.2,
  1.85,
  3.2,
  3.5,
  1.8,
  3.3,
  3.75,
  None,
  None,
  None,
  1.7,
  3.3,
  4.33,
  1.9,
  3.3,
  4,
  1.65,
  3.4,
  4.5,
  1.78,
  3.25,
  4,
  1.73,
  3.4,
  4.2),
 (2,
  1,
  1,
  '2008/2009',
  1,
  '2008-08-16 00:00:00',
  492474,
  10000,
  9994,
  0,
  

In [43]:
c.description

(('id', None, None, None, None, None, None),
 ('country_id', None, None, None, None, None, None),
 ('league_id', None, None, None, None, None, None),
 ('season', None, None, None, None, None, None),
 ('stage', None, None, None, None, None, None),
 ('date', None, None, None, None, None, None),
 ('match_api_id', None, None, None, None, None, None),
 ('home_team_api_id', None, None, None, None, None, None),
 ('away_team_api_id', None, None, None, None, None, None),
 ('home_team_goal', None, None, None, None, None, None),
 ('away_team_goal', None, None, None, None, None, None),
 ('home_player_X1', None, None, None, None, None, None),
 ('home_player_X2', None, None, None, None, None, None),
 ('home_player_X3', None, None, None, None, None, None),
 ('home_player_X4', None, None, None, None, None, None),
 ('home_player_X5', None, None, None, None, None, None),
 ('home_player_X6', None, None, None, None, None, None),
 ('home_player_X7', None, None, None, None, None, None),
 ('home_player_X8', 

1.Which team scored the most points when playing at home?

In [47]:
c.execute('SELECT home_team_api_id, SUM(home_team_goal) FROM Match GROUP BY home_team_api_id ORDER BY SUM(home_team_goal) DESC LIMIT 5')
c.fetchall()

[(8633, 505), (8634, 495), (9925, 389), (9823, 382), (8640, 370)]

Answer: team id 8633.

2.Did this team also score the most points when playing away?

In [48]:
c.execute('SELECT away_team_api_id, SUM(away_team_goal) FROM Match GROUP BY away_team_api_id ORDER BY SUM(away_team_goal) DESC LIMIT 5')
c.fetchall()

[(8634, 354), (8633, 338), (9925, 306), (8593, 287), (8640, 282)]

Answer: No, team id 8634 scored the most points when playing away.

3.How many matches resulted in a tie?

In [52]:
c.execute('SELECT COUNT(id) FROM Match WHERE home_team_goal== away_team_goal')
c.fetchall()

[(6596,)]

4.How many players have Smith for their last name? How many have 'smith' anywhere in their name?

In [62]:
c.execute('SELECT COUNT(*) FROM Player WHERE player_name LIKE "% Smith"')
c.fetchall()

[(15,)]

In [64]:
c.execute('SELECT COUNT(*) FROM Player WHERE player_name LIKE "%Smith%"')
c.fetchall()

[(18,)]

5.What was the median tie score? Use the value determined in the previous question for the number of tie games. Hint: PostgreSQL does not have a median function. Instead, think about the steps required to calculate a median and use the WITH command to store stepwise results as a table and then operate on these results.

In [107]:
c.execute('''WITH tie_table AS (SELECT id, home_team_goal, away_team_goal FROM Match WHERE home_team_goal== away_team_goal ORDER BY home_team_goal DESC) 
SELECT * FROM tie_table LIMIT 1 OFFSET 6596/2''')
c.fetchall()

[(14557, 1, 1)]

In [108]:
c.execute('''WITH tie_table AS (SELECT id, home_team_goal, away_team_goal FROM Match WHERE home_team_goal== away_team_goal ORDER BY home_team_goal DESC) 
SELECT * FROM tie_table LIMIT 1 OFFSET 3299''')
c.fetchall()

[(14561, 1, 1)]

The median is 1.

5.What percentage of players prefer their left or right foot? Hint: Calculate either the right or left foot, whichever is easier based on how you setup the problem.

In [90]:
c.execute('SELECT COUNT(*) FROM Player_Attributes WHERE preferred_foot == "right"')
right = c.fetchall()

In [91]:
c.execute('SELECT COUNT(*) FROM Player_Attributes WHERE preferred_foot == "left"')
left = c.fetchall()

In [96]:
right[0][0] / (right[0][0] + left[0][0])

0.7557469067717946